avoid quadratic array ctor performance; fix #6648

jvasileff opened the pull request:
With patch:

```ceylon
$ ceylon run simple
   ArrayList: 61ms
  LinkedList: 36ms
```

jvasileff committed:
avoid quadratic array ctor performance; fix #6648

gavinking commented:
@jvasileff is this implementation slower for the very common case where the argument is an `ArrayList`?

jvasileff committed:
assuage perf concerns about Array(ArrayList)

jvasileff commented:
Not sure how easy that is to measure, but how about 34398764d ?

jvasileff committed:
Optimize away resizing and copying in Array(List)

jvasileff committed:
don't call elements.getClass() for Array(ArrayList)

In terms of nanoseconds, checking for 'ceylon.collection.ArrayList' is
slow. And, aside from (presumably) reducing megamorphic call sites,
using getFromFirst() doesn't appear to be faster than using an Iterator,
even for size=1 shown below:

1.3.2:
Summary min/max/avg/rstddev/pct
String    43/45/44/1% (100%)
ArrayList 47/55/52/7% (109%)
Range     49/55/53/4% (113%)

This commit:
Summary min/max/avg/rstddev/pct
ArrayList 49/50/49/0% (100%)
String    49/51/50/1% (101%)
Range     52/54/53/1% (106%)

jvasileff committed:
organize code to reduce pull request size

jvasileff commented:
This is done as far as I'm concerned. In the end, it simply uses `iterator.next()` instead of `elements.getFromFirst(i)`.

gavinking commented:
Alright, thanks.
